src/textual/widgets/_selection_list.py

```diff
@@ -0,0 +1,655 @@
+        event.stop()
+        self.post_message(self.SelectionHighlighted(self, event.option_index))
+
+    def _on_option_list_option_selected(self, event: OptionList.OptionSelected) -> None:
+        """Capture the `OptionList` selected event and turn it into a [`SelectionList`][textual.widgets.SelectionList] event.
+
+        Args:
+            event: The event to capture and recreate.
+        """
+        event.stop()
+        self._toggle_highlighted_selection()
+        self.post_message(self.SelectionToggled(self, event.option_index))
+
+    def get_option_at_index(self, index: int) -> Selection[SelectionType]:
+        """Get the selection option at the given index.
+
+        Args:
+            index: The index of the selection option to get.
+
+        Returns:
+            The selection option at that index.
+
+        Raises:
+            OptionDoesNotExist: If there is no selection option with the index.
+        """
+        return cast("Selection[SelectionType]", super().get_option_at_index(index))
+
+    def get_option(self, option_id: str) -> Selection[SelectionType]:
+        """Get the selection option with the given ID.
+
+        Args:
+            index: The ID of the selection option to get.
+
+        Returns:
+            The selection option at with the ID.
+
+        Raises:
+            OptionDoesNotExist: If no selection option has the given ID.
+        """
+        return cast("Selection[SelectionType]", super().get_option(option_id))
+
+    def _remove_option(self, index: int) -> None:
+        """Remove a selection option from the selection option list.
+
+        Args:
+            index: The index of the selection option to remove.
+
+        Raises:
+            IndexError: If there is no selection option of the given index.
+        """
+        self._deselect(self.get_option_at_index(index).value)
+        return super()._remove_option(index)
+
+    def add_options(
+        self,
+        items: Iterable[
+            NewOptionListContent
+            | Selection
+            | tuple[TextType, SelectionType]
+            | tuple[TextType, SelectionType, bool]
+        ],
+    ) -> Self:
+        """Add new selection options to the end of the list.
+
+        Args:
+            items: The new items to add.
+
+        Returns:
+            The [`SelectionList`][textual.widgets.SelectionList] instance.
+
+        Raises:
+            DuplicateID: If there is an attempt to use a duplicate ID.
+            SelectionError: If one of the selection options is of the wrong form.
+        """
+        # This... is sort of sub-optimal, but a natural consequence of
+        # inheriting from and narrowing down OptionList. Here we don't want
+        # things like a separator, or a base Option, being passed in. So we
+        # extend the types of accepted items to keep mypy and friends happy,
+        # but then we runtime check that we've been given sensible types (in
+        # this case the supported tuple values).
```

Comment:
In an ideal world I suspect it would make sense to abstract some of the core of `OptionList` out into something more generic, and then have `OptionList` and `SelectionList` build atop that, yeah.